src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java

```diff
@@ -1320,36 +1338,76 @@ public boolean isReference() {
 
         @Override
         public boolean isPrimitiveClass() {
-            return !isReferenceProjection() && tsym != null && tsym.isPrimitiveClass();
+            // guard against over-eager and/or inopportune completion
+            if (tsym != null) {
+                if (flavor == Flavor.Q_TypeOf_X || tsym.isCompleted()) {
+                    flavor = flavor.metamorphose(tsym.flags());
+                }
+            }
+            return flavor == Flavor.Q_TypeOf_Q || flavor == Flavor.Q_TypeOf_L;
         }
 
         @Override
         public boolean isReferenceProjection() {
-            // gaurd against over-eager and/or inopportune completion
+            // guard against over-eager and/or inopportune completion
+            if (tsym != null) {
+                if (flavor == Flavor.L_TypeOf_X || tsym.isCompleted()) {
+                    flavor = flavor.metamorphose(tsym.flags());
+                }
+            }
+            return flavor == Flavor.L_TypeOf_Q && tsym.type.getFlavor() == Flavor.Q_TypeOf_Q; // discount reference favoring primitives.
+        }
+
+        @Override
+        public boolean isPrimitiveReferenceType() {
+            // guard against over-eager and/or inopportune completion
             if (tsym != null) {
                 if (flavor == Flavor.L_TypeOf_X || tsym.isCompleted()) {
-                    flavor = flavor.metamorphose(tsym.isPrimitiveClass());
+                    flavor = flavor.metamorphose(tsym.flags());
```

Comment:
(Not related to this review) I think we should probably change the `metamorphose` name to a more neutral one like `update` ?

Reply:
Oh, you didn't like the cute name I coined! I really liked metamorphose() since it captures that the Type starts out in larval form where not much is known about and evolves into a full blown form in stages through metamorphosis!

No matter, I am not wedded to the name,  I will add a note to JDK-8268734 to follow up